File: AgentforceSDK-ReactNative-Bridge/android/src/main/java/com/salesforce/android/reactagentforce/BridgeLogger.kt
/*
 * Copyright (c) 2024-present, salesforce.com, inc.
 * All rights reserved.
 *
 * Bridge logger that forwards Agentforce SDK log calls to React Native JS via events.
 */
package com.salesforce.android.reactagentforce

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactContext
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.salesforce.android.mobile.interfaces.logging.Logger

/**
 * Implements the Agentforce SDK Logger interface and forwards log messages
 * to JavaScript via NativeEventEmitter events.
 *
 * Log forwarding is disabled by default and enabled when JS registers a LoggerDelegate.
 */
class BridgeLogger(private val reactContext: ReactContext) : Logger {

    /** Controls whether log messages are forwarded to JS. */
    var forwardingEnabled: Boolean = false

    private fun emit(level: String, message: String, error: String? = null) {
        if (!forwardingEnabled) return

        val params = Arguments.createMap().apply {
            putString("level", level)
            putString("message", message)
            error?.let { putString("error", it) }
        }

        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit("onLogMessage", params)
    }

    override fun e(message: String) = emit("error", message)
    override fun e(message: String, exception: Throwable) = emit("error", message, exception.toString())
    override fun i(message: String) = emit("info", message)
    override fun i(message: String, exception: Throwable) = emit("info", message, exception.toString())
    override fun w(message: String) = emit("warn", message)
    override fun w(message: String, exception: Throwable) = emit("warn", message, exception.toString())
}

File: AgentforceSDK-ReactNative-Bridge/ios/Agentforce/BridgeLogger.swift
//
//  BridgeLogger.swift
//  AgentforceSDK-ReactNative-Bridge
//
//  Created for React Native bridge logging integration
//

import Foundation
import SalesforceLogging

/// Bridge logger that forwards Agentforce SDK logs to React Native JavaScript
///
/// Implements the iOS Agentforce SDK's Logger protocol and emits log events
/// to the React Native event emitter. Forwarding can be enabled/disabled
/// to avoid bridge overhead when no JS delegate is registered.
class BridgeLogger: Logger {

    // MARK: - Properties

    /// Weak reference to the module to avoid retain cycles
    private weak var module: AgentforceModule?

    private let lock = NSLock()
    private var _forwardingEnabled: Bool = false

    /// Controls whether logs should be forwarded to JavaScript.
    /// Thread-safe: written on the JS thread, read on SDK callback threads.
    var forwardingEnabled: Bool {
        get { lock.withLock { _forwardingEnabled } }
        set { lock.withLock { _forwardingEnabled = newValue } }
    }

    // MARK: - Initialization

    init(module: AgentforceModule) {
        self.module = module
    }

    // MARK: - Logger Protocol Implementation

    /// Logs a message at the specified level
    /// Forwards to React Native if forwarding is enabled
    func log(_ logMessage: String, level: LogLevel) {
        guard forwardingEnabled else { return }

        // Map iOS LogLevel to TypeScript string literal
        let levelString: String
        switch level {
        case .error:
            levelString = "error"
        case .warning:
            levelString = "warn"
        case .info:
            levelString = "info"
        case .debug:
            levelString = "debug"
        @unknown default:
            levelString = "info"  // Fallback for future log levels
        }

        // Build event payload matching TypeScript LoggerDelegate interface
        let payload: [String: Any] = [
            "level": levelString,
            "message": logMessage
        ]

        // Emit to React Native (must be called from module with hasListeners check)
        module?.emitLogEvent(payload)
    }
}
